homeassistant/helpers/event.py

```diff
@@ -530,29 +607,37 @@ def calculate_next(now: datetime) -> None:
 
     # Make sure rolling back the clock doesn't prevent the timer from
     # triggering.
-    last_now: Optional[datetime] = None
+    cancel_callback: Optional[asyncio.TimerHandle] = None
+    calculate_next(next_time)
 
     @callback
-    def pattern_time_change_listener(event: Event) -> None:
+    def pattern_time_change_listener() -> None:
         """Listen for matching time_changed events."""
-        nonlocal next_time, last_now
+        nonlocal next_time, cancel_callback
 
-        now = event.data[ATTR_NOW]
+        now = pattern_utc_now()
+        hass.async_run_job(action, dt_util.as_local(now) if local else now)
 
-        if last_now is None or now < last_now:
-            # Time rolled back or next time not yet calculated
-            calculate_next(now)
+        calculate_next(now + timedelta(seconds=1))
 
-        last_now = now
+        cancel_callback = hass.loop.call_at(
+            hass.loop.time() + next_time.timestamp() - time.time(),
+            pattern_time_change_listener,
+        )
 
-        if next_time <= now:
-            hass.async_run_job(action, dt_util.as_local(now) if local else now)
-            calculate_next(now + timedelta(seconds=1))
+    cancel_callback = hass.loop.call_at(
+        hass.loop.time() + next_time.timestamp() - time.time(),
+        pattern_time_change_listener,
+    )
+
+    @callback
+    def unsub_pattern_time_change_listener() -> None:
+        """Cancel the call_later."""
+        nonlocal cancel_callback
+        assert cancel_callback is not None
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5872960)